File: clients/javascript/src/resources/contracts.ts
import type { HieroProxyClient } from "../client";
import type {
  ContractCreateRequest,
  ContractCallRequest,
  ContractResponse,
  ContractCallResultResponse,
} from "./contracts.types";
import { requireId } from "./validation";

/**
 * Smart Contracts resource — deploy, call, and query contracts.
 */
export class ContractsResource {
  constructor(private readonly client: HieroProxyClient) {}

  /**
   * Deploy a smart contract.
   * @param bytecode Hex-encoded contract bytecode.
   */
  async deploy(bytecode: string): Promise<ContractResponse> {
    const body: ContractCreateRequest = { bytecode };
    return this.client.post("/api/v1/contracts", body);
  }

  /**
   * Call a smart contract function.
   */
  async call(contractId: string, functionName: string): Promise<ContractCallResultResponse> {
    const id = requireId(contractId, "contractId");
    const body: ContractCallRequest = { functionName };
    return this.client.post(`/api/v1/contracts/${id}/call`, body);
  }

  /**
   * List all contracts from the mirror node.
   */
  async list(): Promise<ContractResponse[]> {
    return this.client.get("/api/v1/contracts");
  }

  /**
   * Get contract info by ID.
   */
  async getInfo(contractId: string): Promise<ContractResponse> {
    const id = requireId(contractId, "contractId");
    return this.client.get(`/api/v1/contracts/${id}`);
  }
}

File: clients/python/src/hiero_proxy/resources/contracts.py
"""Contracts resource."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from hiero_proxy.resources.validation import require_id

if TYPE_CHECKING:
    from hiero_proxy.client import AsyncHieroProxyClient, HieroProxyClient


class ContractsResource:
    """Smart contract operations — deploy, call, query."""

    def __init__(self, client: HieroProxyClient | AsyncHieroProxyClient) -> None:
        self._client = client

    def deploy(self, bytecode: str) -> Any:
        return self._client.post("/api/v1/contracts", json={"bytecode": bytecode})

    def call(self, contract_id: str, function_name: str) -> Any:
        cid = require_id(contract_id, "contract_id")
        return self._client.post(f"/api/v1/contracts/{cid}/call", json={"functionName": function_name})

    def list(self) -> Any:
        return self._client.get("/api/v1/contracts")

    def get_info(self, contract_id: str) -> Any:
        cid = require_id(contract_id, "contract_id")
        return self._client.get(f"/api/v1/contracts/{cid}")
